File: kittens/diff/collect.py
#!/usr/bin/env python3
# vim:fileencoding=utf-8
# License: GPL v3 Copyright: 2018, Kovid Goyal <kovid at kovidgoyal.net>

import os
import re
from contextlib import suppress
from functools import lru_cache
from hashlib import md5
from kitty.guess_mime_type import guess_type
from typing import TYPE_CHECKING, Dict, List, Set, Optional, Iterator, Tuple, Union

if TYPE_CHECKING:
    from .highlight import DiffHighlight  # noqa


path_name_map: Dict[str, str] = {}
remote_dirs: Dict[str, str] = {}


def add_remote_dir(val: str) -> None:
    remote_dirs[val] = os.path.basename(val).rpartition('-')[-1]


class Segment:

    __slots__ = ('start', 'end', 'start_code', 'end_code')

    def __init__(self, start: int, start_code: str):
        self.start = start
        self.start_code = start_code
        self.end: Optional[int] = None
        self.end_code: Optional[str] = None

    def __repr__(self) -> str:
        return 'Segment(start={!r}, start_code={!r}, end={!r}, end_code={!r})'.format(
            self.start, self.start_code, self.end, self.end_code)


class Collection:

    def __init__(self) -> None:
        self.changes: Dict[str, str] = {}
        self.renames: Dict[str, str] = {}
        self.adds: Set[str] = set()
        self.removes: Set[str] = set()
        self.all_paths: List[str] = []
        self.type_map: Dict[str, str] = {}
        self.added_count = self.removed_count = 0

    def add_change(self, left_path: str, right_path: str) -> None:
        self.changes[left_path] = right_path
        self.all_paths.append(left_path)
        self.type_map[left_path] = 'diff'

    def add_rename(self, left_path: str, right_path: str) -> None:
        self.renames[left_path] = right_path
        self.all_paths.append(left_path)
        self.type_map[left_path] = 'rename'

    def add_add(self, right_path: str) -> None:
        self.adds.add(right_path)
        self.all_paths.append(right_path)
        self.type_map[right_path] = 'add'
        if isinstance(data_for_path(right_path), str):
            self.added_count += len(lines_for_path(right_path))

    def add_removal(self, left_path: str) -> None:
        self.removes.add(left_path)
        self.all_paths.append(left_path)
        self.type_map[left_path] = 'removal'
        if isinstance(data_for_path(left_path), str):
            self.removed_count += len(lines_for_path(left_path))

    def finalize(self) -> None:
        def key(x: str) -> str:
            return path_name_map.get(x, '')

        self.all_paths.sort(key=key)

    def __iter__(self) -> Iterator[Tuple[str, str, Optional[str]]]:
        for path in self.all_paths:
            typ = self.type_map[path]
            if typ == 'diff':
                data: Optional[str] = self.changes[path]
            elif typ == 'rename':
                data = self.renames[path]
            else:
                data = None
            yield path, typ, data

    def __len__(self) -> int:
        return len(self.all_paths)


def remote_hostname(path: str) -> Tuple[Optional[str], Optional[str]]:
    for q in remote_dirs:
        if path.startswith(q):
            return q, remote_dirs[q]
    return None, None


def resolve_remote_name(path: str, default: str) -> str:
    remote_dir, rh = remote_hostname(path)
    if remote_dir and rh:
        return rh + ':' + os.path.relpath(path, remote_dir)
    return default


def collect_files(collection: Collection, left: str, right: str) -> None:
    left_names: Set[str] = set()
    right_names: Set[str] = set()
    left_path_map: Dict[str, str] = {}
    right_path_map: Dict[str, str] = {}

    def walk(base: str, names: Set[str], pmap: Dict[str, str]) -> None:
        for dirpath, dirnames, filenames in os.walk(base):
            for filename in filenames:
                path = os.path.abspath(os.path.join(dirpath, filename))
                path_name_map[path] = name = os.path.relpath(path, base)
                names.add(name)
                pmap[name] = path

    walk(left, left_names, left_path_map)
    walk(right, right_names, right_path_map)
    common_names = left_names & right_names
    changed_names = {n for n in common_names if data_for_path(left_path_map[n]) != data_for_path(right_path_map[n])}
    for n in changed_names:
        collection.add_change(left_path_map[n], right_path_map[n])

    removed = left_names - common_names
    added = right_names - common_names
    ahash = {a: hash_for_path(right_path_map[a]) for a in added}
    rhash = {r: hash_for_path(left_path_map[r]) for r in removed}
    for name, rh in rhash.items():
        for n, ah in ahash.items():
            if ah == rh and data_for_path(left_path_map[name]) == data_for_path(right_path_map[n]):
                collection.add_rename(left_path_map[name], right_path_map[n])
                added.discard(n)
                break
        else:
            collection.add_removal(left_path_map[name])

    for name in added:
        collection.add_add(right_path_map[name])


sanitize_pat = re.compile('[\x00-\x09\x0b-\x1f\x7f\x80-\x9f]')


def sanitize(text: str) -> str:
    ntext = text.replace('\r\n', '⏎\n')
    return sanitize_pat.sub('░', ntext)


@lru_cache(maxsize=1024)
def mime_type_for_path(path: str) -> str:
    return guess_type(path) or 'application/octet-stream'


@lru_cache(maxsize=1024)
def raw_data_for_path(path: str) -> bytes:
    with open(path, 'rb') as f:
        return f.read()


def is_image(path: Optional[str]) -> bool:
    return mime_type_for_path(path).startswith('image/') if path else False


@lru_cache(maxsize=1024)
def data_for_path(path: str) -> Union[str, bytes]:
    raw_bytes = raw_data_for_path(path)
    if not is_image(path) and not os.path.samefile(path, os.devnull):
        with suppress(UnicodeDecodeError):
            return raw_bytes.decode('utf-8')
    return raw_bytes


class LinesForPath:

    replace_tab_by = ' ' * 4

    @lru_cache(maxsize=1024)
    def __call__(self, path: str) -> Tuple[str, ...]:
        data = data_for_path(path)
        assert isinstance(data, str)
        data = data.replace('\t', self.replace_tab_by)
        return tuple(sanitize(data).splitlines())


lines_for_path = LinesForPath()


@lru_cache(maxsize=1024)
def hash_for_path(path: str) -> bytes:
    return md5(raw_data_for_path(path)).digest()


def create_collection(left: str, right: str) -> Collection:
    collection = Collection()
    if os.path.isdir(left):
        collect_files(collection, left, right)
    else:
        pl, pr = os.path.abspath(left), os.path.abspath(right)
        path_name_map[pl] = resolve_remote_name(pl, left)
        path_name_map[pr] = resolve_remote_name(pr, right)
        collection.add_change(pl, pr)
    collection.finalize()
    return collection


highlight_data: Dict[str, 'DiffHighlight'] = {}


def set_highlight_data(data: Dict[str, 'DiffHighlight']) -> None:
    global highlight_data
    highlight_data = data


def highlights_for_path(path: str) -> 'DiffHighlight':
    return highlight_data.get(path, [])

File: kittens/diff/collect.go
// License: GPLv3 Copyright: 2023, Kovid Goyal, <kovid at kovidgoyal.net>

package diff

import (
	"crypto/md5"
	"fmt"
	"io/fs"
	"os"
	"path/filepath"
	"strings"
	"unicode/utf8"

	"github.com/kovidgoyal/kitty/tools/utils"
)

var _ = fmt.Print
var path_name_map, remote_dirs map[string]string

var mimetypes_cache, data_cache, hash_cache *utils.LRUCache[string, string]
var size_cache *utils.LRUCache[string, int64]
var lines_cache *utils.LRUCache[string, []string]
var light_highlighted_lines_cache *utils.LRUCache[string, []string]
var dark_highlighted_lines_cache *utils.LRUCache[string, []string]
var is_text_cache *utils.LRUCache[string, bool]

func init_caches() {
	path_name_map = make(map[string]string, 32)
	remote_dirs = make(map[string]string, 32)
	const sz = 4096
	size_cache = utils.NewLRUCache[string, int64](sz)
	mimetypes_cache = utils.NewLRUCache[string, string](sz)
	data_cache = utils.NewLRUCache[string, string](sz)
	is_text_cache = utils.NewLRUCache[string, bool](sz)
	lines_cache = utils.NewLRUCache[string, []string](sz)
	light_highlighted_lines_cache = utils.NewLRUCache[string, []string](sz)
	dark_highlighted_lines_cache = utils.NewLRUCache[string, []string](sz)
	hash_cache = utils.NewLRUCache[string, string](sz)
}

func add_remote_dir(val string) {
	x := filepath.Base(val)
	idx := strings.LastIndex(x, "-")
	if idx > -1 {
		x = x[idx+1:]
	} else {
		x = ""
	}
	remote_dirs[val] = x
}

func mimetype_for_path(path string) string {
	return mimetypes_cache.MustGetOrCreate(path, func(path string) string {
		mt := utils.GuessMimeTypeWithFileSystemAccess(path)
		if mt == "" {
			mt = "application/octet-stream"
		}
		if utils.KnownTextualMimes[mt] {
			if _, a, found := strings.Cut(mt, "/"); found {
				mt = "text/" + a
			}
		}
		return mt
	})
}

func data_for_path(path string) (string, error) {
	return data_cache.GetOrCreate(path, func(path string) (string, error) {
		ans, err := os.ReadFile(path)
		return utils.UnsafeBytesToString(ans), err
	})
}

func size_for_path(path string) (int64, error) {
	return size_cache.GetOrCreate(path, func(path string) (int64, error) {
		s, err := os.Stat(path)
		if err != nil {
			return 0, err
		}
		return s.Size(), nil
	})
}

func is_image(path string) bool {
	return strings.HasPrefix(mimetype_for_path(path), "image/")
}

func is_path_text(path string) bool {
	return is_text_cache.MustGetOrCreate(path, func(path string) bool {
		if is_image(path) {
			return false
		}
		s1, err := os.Stat(path)
		if err == nil {
			s2, err := os.Stat("/dev/null")
			if err == nil && os.SameFile(s1, s2) {
				return false
			}
		}
		d, err := data_for_path(path)
		if err != nil {
			return false
		}
		return utf8.ValidString(d)
	})
}

func hash_for_path(path string) (string, error) {
	return hash_cache.GetOrCreate(path, func(path string) (string, error) {
		ans, err := data_for_path(path)
		if err != nil {
			return "", err
		}
		hash := md5.Sum(utils.UnsafeStringToBytes(ans))
		return utils.UnsafeBytesToString(hash[:]), err
	})

}

func text_to_lines(text string) []string {
	lines := make([]string, 0, 512)
	splitlines_like_git(text, false, func(line string) { lines = append(lines, line) })
	return lines
}

func sanitize(text string) string { return utils.ReplaceControlCodes(text, conf.Replace_tab_by, "\n") }

func lines_for_path(path string) ([]string, error) {
	return lines_cache.GetOrCreate(path, func(path string) ([]string, error) {
		ans, err := data_for_path(path)
		if err != nil {
			return nil, err
		}
		return text_to_lines(sanitize(ans)), nil
	})
}

func highlighted_lines_for_path(path string) ([]string, error) {
	plain_lines, err := lines_for_path(path)
	if err != nil {
		return nil, err
	}
	var ans []string
	var found bool
	if use_light_colors {
		ans, found = light_highlighted_lines_cache.Get(path)
	} else {
		ans, found = dark_highlighted_lines_cache.Get(path)
	}
	if found && len(ans) == len(plain_lines) {
		return ans, nil
	}
	return plain_lines, nil
}

type Collection struct {
	changes, renames, type_map map[string]string
	adds, removes              *utils.Set[string]
	all_paths                  []string
	paths_to_highlight         *utils.Set[string]
	added_count, removed_count int
}

func (self *Collection) add_change(left, right string) {
	self.changes[left] = right
	self.all_paths = append(self.all_paths, left)
	self.paths_to_highlight.Add(left)
	self.paths_to_highlight.Add(right)
	self.type_map[left] = `diff`
}

func (self *Collection) add_rename(left, right string) {
	self.renames[left] = right
	self.all_paths = append(self.all_paths, left)
	self.type_map[left] = `rename`
}

func (self *Collection) add_add(right string) {
	self.adds.Add(right)
	self.all_paths = append(self.all_paths, right)
	self.paths_to_highlight.Add(right)
	self.type_map[right] = `add`
	if is_path_text(right) {
		num, _ := lines_for_path(right)
		self.added_count += len(num)
	}
}

func (self *Collection) add_removal(left string) {
	self.removes.Add(left)
	self.all_paths = append(self.all_paths, left)
	self.paths_to_highlight.Add(left)
	self.type_map[left] = `removal`
	if is_path_text(left) {
		num, _ := lines_for_path(left)
		self.removed_count += len(num)
	}
}

func (self *Collection) finalize() {
	utils.StableSortWithKey(self.all_paths, func(path string) string {
		return path_name_map[path]
	})
}

func (self *Collection) Len() int { return len(self.all_paths) }

func (self *Collection) Items() int { return len(self.all_paths) }

func (self *Collection) Apply(f func(path, typ, changed_path string) error) error {
	for _, path := range self.all_paths {
		typ := self.type_map[path]
		changed_path := ""
		switch typ {
		case "diff":
			changed_path = self.changes[path]
		case "rename":
			changed_path = self.renames[path]
		}
		if err := f(path, typ, changed_path); err != nil {
			return err
		}
	}
	return nil
}

func allowed(path string, patterns ...string) bool {
	name := filepath.Base(path)
	for _, pat := range patterns {
		if matched, err := filepath.Match(pat, name); err == nil && matched {
			return false
		}
	}
	return true
}

func remote_hostname(path string) (string, string) {
	for q, val := range remote_dirs {
		if strings.HasPrefix(path, q) {
			return q, val
		}
	}
	return "", ""
}

func resolve_remote_name(path, defval string) string {
	remote_dir, rh := remote_hostname(path)
	if remote_dir != "" && rh != "" {
		r, err := filepath.Rel(remote_dir, path)
		if err == nil {
			return rh + ":" + r
		}
	}
	return defval
}

func walk(base string, patterns []string, names *utils.Set[string], pmap, path_name_map map[string]string) error {
	base, err := filepath.Abs(base)
	if err != nil {
		return err
	}
	return filepath.WalkDir(base, func(path string, d fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		is_allowed := allowed(path, patterns...)
		if !is_allowed {
			if d.IsDir() {
				return fs.SkipDir
			}
			return nil
		}
		if d.IsDir() {
			return nil
		}
		path, err = filepath.Abs(path)
		if err != nil {
			return err
		}
		name, err := filepath.Rel(base, path)
		if err != nil {
			return err
		}
		if name != "." {
			path_name_map[path] = name
			names.Add(name)
			pmap[name] = path
		}
		return nil
	})
}

func (self *Collection) collect_files(left, right string) error {
	left_names, right_names := utils.NewSet[string](16), utils.NewSet[string](16)
	left_path_map, right_path_map := make(map[string]string, 16), make(map[string]string, 16)
	err := walk(left, conf.Ignore_name, left_names, left_path_map, path_name_map)
	if err != nil {
		return err
	}
	if err = walk(right, conf.Ignore_name, right_names, right_path_map, path_name_map); err != nil {
		return err
	}
	common_names := left_names.Intersect(right_names)
	changed_names := utils.NewSet[string](common_names.Len())
	for n := range common_names.Iterable() {
		ld, err := data_for_path(left_path_map[n])
		var rd string
		if err == nil {
			rd, err = data_for_path(right_path_map[n])
		}
		if err != nil {
			return err
		}
		if ld != rd {
			changed_names.Add(n)
			self.add_change(left_path_map[n], right_path_map[n])
		} else {
			if lstat, err := os.Stat(left_path_map[n]); err == nil {
				if rstat, err := os.Stat(right_path_map[n]); err == nil {
					if lstat.Mode() != rstat.Mode() {
						// identical files with only a mode change
						changed_names.Add(n)
						self.add_change(left_path_map[n], right_path_map[n])
					}
				}
			}
		}
	}
	removed := left_names.Subtract(common_names)
	added := right_names.Subtract(common_names)
	ahash, rhash := make(map[string]string, added.Len()), make(map[string]string, removed.Len())
	for a := range added.Iterable() {
		ahash[a], err = hash_for_path(right_path_map[a])
		if err != nil {
			return err
		}
	}
	for r := range removed.Iterable() {
		rhash[r], err = hash_for_path(left_path_map[r])
		if err != nil {
			return err
		}
	}
	for name, rh := range rhash {
		found := false
		for n, ah := range ahash {
			if ah == rh {
				ld, _ := data_for_path(left_path_map[name])
				rd, _ := data_for_path(right_path_map[n])
				if ld == rd {
					self.add_rename(left_path_map[name], right_path_map[n])
					added.Discard(n)
					found = true
					break
				}
			}
		}
		if !found {
			self.add_removal(left_path_map[name])
		}
	}
	for name := range added.Iterable() {
		self.add_add(right_path_map[name])
	}
	return nil
}

func create_collection(left, right string) (ans *Collection, err error) {
	ans = &Collection{
		changes:            make(map[string]string),
		renames:            make(map[string]string),
		type_map:           make(map[string]string),
		adds:               utils.NewSet[string](32),
		removes:            utils.NewSet[string](32),
		paths_to_highlight: utils.NewSet[string](32),
		all_paths:          make([]string, 0, 32),
	}
	left_stat, err := os.Stat(left)
	if err != nil {
		return nil, err
	}
	if left_stat.IsDir() {
		err = ans.collect_files(left, right)
		if err != nil {
			return nil, err
		}
	} else {
		pl, err := filepath.Abs(left)
		if err != nil {
			return nil, err
		}
		pr, err := filepath.Abs(right)
		if err != nil {
			return nil, err
		}
		path_name_map[pl] = resolve_remote_name(pl, left)
		path_name_map[pr] = resolve_remote_name(pr, right)
		ans.add_change(pl, pr)
	}
	ans.finalize()
	return ans, err
}
